### PR TITLE
Change to streaming out the heap snapshot data

### PR DESCRIPTION
**Pull request by JianFangAtRai**
This PR is to continue the work on the following PR: 

Prevent OOMs during heap snapshot: Change to streaming out the snapshot data (https://github.com/JuliaLang/julia/pull/51518 )

Here are the commit history:

```
* Streaming the heap snapshot!

This should prevent the engine from OOMing while recording the snapshot!

Now we just need to sample the files, either online, before downloading, or offline after downloading :)

If we're gonna do it offline, we'll want to gzip the files before downloading them.

* Allow custom filename; use original API

* Support legacy heap snapshot interface. Add reassembly function.

* Add tests

* Apply suggestions from code review

* Update src/gc-heap-snapshot.cpp

* Change to always save the parts in the same directory

This way you can always recover from an OOM

* Fix bug in reassembler: from_node and to_node were in the wrong order

* Fix correctness mistake: The edges have to be reordered according to the node order. That's the whole reason this is tricky.

But i'm not sure now whether the SoAs approach is actually an optimization.... It seems like we should probably prefer to inline the Edges right into the vector, rather than having to do another random lookup into the edges table?

* Debugging messed up edge array idxs

* Disable log message

* Write the .nodes and .edges as binary data

* Remove unnecessary logging

* fix merge issues

* attempt to add back the orphan node checking logic
```

### CODE REVIEW

**Comment by inkydragon** [comment]
Since the current changes do not involve documentation, CI errors are irrelevant.
You may want to rebase based on the master branch.

```
┌ Error: no doc found for reference '[`Profile.HeapSnapshot.assemble_snapshot(filepath)`](@ref)' in src\stdlib\Profile.md.
└ @ Documenter C:\workdir\doc\deps\packages\Documenter\1HwWe\src\utilities\utilities.jl:44
[ Info: CheckDocument: running document checks.
[ Info: Populate: populating indices.
ERROR: LoadError: `makedocs` encountered an error [:cross_references] -- terminating build before rendering.
```

**Comment by gbaraldi** [comment]
I'm not super familiar with this code, I've only modified it a bit once, but the changes seems pretty reasonable. Did you folks do some testing to see if chrome is still able to process the file that gets regenerated from the json?

**Comment by JianFangAtRai** [comment]
> I'm not super familiar with this code, I've only modified it a bit once, but the changes seems pretty reasonable. Did you folks do some testing to see if chrome is still able to process the file that gets regenerated from the json?

Yes, we tested the snapshot with different sizes from hundreds of MBs to couple GBs, chrome was still able to process the assembled files.

File: src/gc-heap-snapshot.cpp [hunk 1]
```diff
@@ -10,12 +10,14 @@
 #include "llvm/ADT/StringMap.h"
 #include "llvm/ADT/DenseMap.h"
 
+#include <atomic>
```
[the remark this answers is not in the document]
Comment (JianFangAtRai) [reply src/gc-heap-snapshot.cpp]: make sense, removed atomic to use integer directly. thanks

File: stdlib/Profile/src/heapsnapshot_reassemble.jl [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+Assemble a .heapsnapshot file from the .json files produced by `Profile.take_snapshot`.
+"""
+
+# SoA layout to reduce padding
+struct Edges
+    type::Vector{Int8}       # index into `snapshot.meta.edge_types`
+    name_or_index::Vector{UInt} # Either an index into `snapshot.strings`, or the index in an array, depending on edge_type
+    to_pos::Vector{UInt}   # index into `snapshot.nodes`
+end
+function init_edges(n::Int)
```
[the remark this answers is not in the document]
Comment (JianFangAtRai) [reply stdlib/Profile/src/heapsnapshot_reassemble.jl]: updated as suggested. thanks

File: stdlib/Profile/src/heapsnapshot_reassemble.jl [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    edge_count::Vector{UInt} # number of outgoing edges
+    edges::Edges               # outgoing edges
+    # This is the main complexity of the .heapsnapshot format, and it's the reason we need
+    # to read in all the data before writing it out. The edges vector contains all edges,
+    # but organized by which node they came from. First, it contains all the edges coming
+    # out of node 0, then all edges leaving node 1, etc. So we need to have visited all
+    # edges, and assigned them to their corresponding nodes, before we can emit the file.
+    edge_idxs::Vector{Vector{UInt}} # indexes into edges, keeping per-node outgoing edge ids
+end
+function init_nodes(n::Int, e::Int)
```
[the remark this answers is not in the document]
Comment (JianFangAtRai) [reply stdlib/Profile/src/heapsnapshot_reassemble.jl]: updated.

**Pull request merged**